posts/2020-07-16-Rust-1.45.0.md

```diff
@@ -0,0 +1,296 @@
+
+```rust
+fn cast(x: f32) -> u8 {
+    x as u8
+}
+
+fn main() {
+    let too_big = 300.0;
+    let too_small = -100.0;
+    let nan = f32::NAN;
+    
+    println!("too_big_casted = {}", cast(too_big));
+    println!("too_small_casted = {}", cast(too_small));
+    println!("not_a_number_casted = {}", cast(nan));
+}
+```
+
+This will print:
+
+```text
+too_big_casted = 255
+too_small_casted = 0
+not_a_number_casted = 0
+```
+
+That is, numbers that are too big turn into the largest possible value.
+Numbers that are too small produce the smallest possible value (which is
+zero). NaN produces zero.
+
+The new API to cast in an unsafe manner is:
+
+```rust
+let x: f32 = 1.0;
+let y: u8 = unsafe { x.to_int_unchecked() };
+```
+
+But as always, you should only use this method as a last resort. Just like
+with array access, the compiler can often optimize the checks away, making
+the safe and unsafe versions equivalent when the compiler can prove it.
+
+## Stabilizing function-like procedural macros in expressions, patterns, and statements
+
+In [Rust 1.30.0](https://blog.rust-lang.org/2018/10/25/Rust-1.30.0.html), we stabilized
+"function-like procedural macros in item position." For example, [the
+`gnome-class` crate](https://gitlab.gnome.org/federico/gnome-class):
+
+> Gnome-class is a procedural macro for Rust.  Within the macro, we
+> define a mini-language which looks as Rust-y as possible, and that has
+> extensions to let you define GObject subclasses, their properties,
+> signals, interface implementations, and the rest of GObject's
+> features.  The goal is to require no unsafe code on your part.
+
+This looks like this:
+
+```rust
+gobject_gen! {
+    class MyClass: GObject {
+        foo: Cell<i32>,
+        bar: RefCell<String>,
+    }
+
+    impl MyClass {
+        virtual fn my_virtual_method(&self, x: i32) {
+            ... do something with x ...
+        }
+    }
+}
+```
+
+The "in item position" bit is some jargon, but basically what this means is that
+you could only invoke `gobject_gen!` in certain places in your code.
+
+Rust 1.45.0 adds the ability to invoke procedural macros in three new places:
+
+```rust
+// imagine we have a procedural macro named "mac"
+
+mac!(); // item position, this was what was stable before
+
+// but these three are new:
+fn main() {
+  let expr = mac!(); // expression position
+
+  match expr {
+      mac!() => {} // pattern position
+  }
+
+  mac!(); // statement position
+}
+```
+
+Being able to use macros in more places is interesting, but there's another
+reason why many Rustaceans have been waiting for this feature for a long
+time. [Rocket](https://rocket.rs/) is a popular web framework for Rust.
+Initially released in December of 2016, it has often been described as one of
+the best things the Rust ecosystem has to offer. Here's the "hello world"
+example from its homepage:
+
+```rust
+#![feature(proc_macro_hygiene, decl_macro)]
```

Comment:
Would it be better to use the expected example from `0.5`?

```rust
#[macro_use] extern crate rocket;

#[get("/<name>/<age>")]
fn hello(name: String, age: u8) -> String {
    format!("Hello, {} year old named {}!", age, name)
}

#[launch]
fn rocket() -> rocket::Rocket {
    rocket::ignite().mount("/hello", routes![hello])
}
```

If so, perhaps the text above could change to:

> Being able to use macros in more places is interesting, but there's another
reason why many Rustaceans have been waiting for this feature for a long time:
[Rocket](https://rocket.rs/). Inititally released in December of 2016, Rocket is
a popular web framework for Rust often described as one of the best things the
Rust ecosystem has to offer. Here's the "hello world" example from its upcoming release:

Reply:
@SergioBenitez done!

Reply:
The next paragraph will also need to change to accommodate this change. Drafting a suggestion now.